src/lib/upstream_requests_manager.ts

```diff
@@ -0,0 +1,67 @@
+import got from 'got'
+
+// TODO: Consider changing this to:
+// export type UpstreamResponse<Data> = {
+//   data: Data
+//   headers: Headers  // from @mapbox/mbtiles
+// }
+export type UpstreamResponse<Data> = {
+  data: Data
+  etag?: string
+}
```

Comment:
```suggestion
type ResponseType = 'buffer' | 'json' | 'text'
type DataType<T extends ResponseType> = T extends 'buffer' ? Buffer : T extends 'json' ? TileJSON : T extends 'text' ? string : never

export type UpstreamResponse<T extends ResponseType> = {
  data: DataType<T>
  etag?: string
}
```

Reply:
mind blown! couldn't figure this out for the life of me 😅